src/fdopendir.c

```diff
@@ -49,45 +78,45 @@ DIR *fdopendir(int dirfd) {
         return 0;
     }
 
-    if (dirfd == AT_FDCWD) {
-        dir = opendir (".");
-        /* dirfd can be closed only upon success */
-        if (dir) PROTECT_ERRNO(close(dirfd));
-        return dir;
-    }
+    /* Remember the current directory for later */
 
     oldCWD = open(".", O_RDONLY);
     if (oldCWD == -1)
         return 0;
 
-    if(best_fchdir(dirfd) < 0) {
-        if (oldCWD != -1) PROTECT_ERRNO(close(oldCWD));
+    /* Change to the supplied directory, open it, then change back */
+
+    if (fchdir(dirfd) < 0) {
+        PROTECT_ERRNO(close(oldCWD));
         return 0;
     }
 
-    dir = opendir (".");
+    dirvp = (void *)(opendir)(".");
 
-    if (best_fchdir(oldCWD) < 0) {
-        if (dir) PROTECT_ERRNO(closedir(dir));
-        if (oldCWD != -1) PROTECT_ERRNO(close(oldCWD));
+    if (fchdir(oldCWD) < 0) {
```

Comment:
I think this needs to stay as `best_fchdir()`, in order to call `__pthread_fchdir()` if possible.  Otherwise it's no longer thread safe.

Reply:
Actually I think I see why this wasn't working for you, the current (before this patch) `fdopendir()` looks broken to me.  `best_fchdir()` has to be used in [the way that `_ATCALL()` uses it](https://github.com/macports/macports-legacy-support/blob/b570b7f97476e5377ee2397a7a3a71638b72bdb5/src/common-priv.h#L69).  I was going to write a long comment describing how that seems to work, but then I realized you can probably just do this in `fdopendir()`:

```c
DIR *dirp = _ATCALL(dirfd, ".", NULL, opendir("."));
```

That way you don't have to track `oldCWD` yourself or anything.  Should simplify it a lot.

Reply:
Thanks. That's perfect. It fixes the bug and removes code duplication. It's done.